ec/src/models/mnt6/g2.rs

```diff
@@ -53,32 +55,28 @@ impl<P: MNT6Parameters> From<G2Affine<P>> for G2Prepared<P> {
             t: <Fp3<P::Fp3Config>>::one(),
         };
 
-        for (idx, value) in P::ATE_LOOP_COUNT.iter().rev().enumerate() {
-            let mut tmp = *value;
-            let skip_extraneous_bits = 64 - value.leading_zeros();
-            let mut v = Vec::with_capacity(16);
-            for i in 0..64 {
-                if idx == 0 && (i == 0 || i >= skip_extraneous_bits) {
-                    continue;
-                }
-                v.push(tmp & 1 == 1);
-                tmp >>= 1;
-            }
-
-            for bit in v.iter().rev() {
-                let (r2, coeff) = MNT6::<P>::doubling_step_for_flipped_miller_loop(&r);
-                g2p.double_coefficients.push(coeff);
-                r = r2;
-
-                if *bit {
-                    let (r2, coeff) =
+        let neg_g2 = g2.neg();
+        for bit in P::ATE_LOOP_COUNT.iter().skip(1) {
+            let (r2, coeff) = MNT6::<P>::doubling_step_for_flipped_miller_loop(&r);
+            g2p.double_coefficients.push(coeff);
+            r = r2;
+
+            let add_coeff;
+            let r_temp;
+            match bit {
+                1 => {
+                    (r_temp, add_coeff) =
                         MNT6::<P>::mixed_addition_step_for_flipped_miller_loop(&g2.x, &g2.y, &r);
-                    g2p.addition_coefficients.push(coeff);
-                    r = r2;
-                }
-
-                tmp >>= 1;
+                },
+                -1 => {
+                    (r_temp, add_coeff) = MNT6::<P>::mixed_addition_step_for_flipped_miller_loop(
+                        &neg_g2.x, &neg_g2.y, &r,
+                    );
+                },
+                _ => continue,
```

Comment:
Some potential change
```suggestion
                0 => continue,
                _ => unimplemented!()
```